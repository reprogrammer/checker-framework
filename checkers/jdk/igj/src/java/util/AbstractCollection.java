package java.util;

import checkers.igj.quals.*;

@I
public abstract class AbstractCollection<E> implements Collection<E> {
  protected AbstractCollection() @ReadOnly {}
  public abstract @I Iterator<E> iterator() @ReadOnly;
  public abstract int size() @ReadOnly;
  public boolean isEmpty() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean contains(Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public Object[] toArray() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public <T> T @Mutable [] toArray(T @Mutable [] a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean add(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean remove(Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean containsAll(@ReadOnly Collection<?> a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean addAll(@ReadOnly Collection<? extends E> a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean removeAll(@ReadOnly Collection<?> a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean retainAll(@ReadOnly Collection<?> a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public String toString() @ReadOnly { throw new RuntimeException("skeleton method"); }
}
