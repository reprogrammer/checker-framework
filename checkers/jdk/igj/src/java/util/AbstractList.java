package java.util;

import checkers.igj.quals.*;

@I
public abstract class AbstractList<E> extends @I AbstractCollection<E> implements @I List<E> {
  protected AbstractList() @ReadOnly {}
  public boolean add(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public abstract E get(int a1) @ReadOnly;
  public E set(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public void add(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public E remove(int a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public int indexOf(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int lastIndexOf(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean addAll(int a1, @ReadOnly Collection<? extends E> a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public @I Iterator<E> iterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I ListIterator<E> listIterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I ListIterator<E> listIterator(int a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I List<E> subList(int a1, int a2) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean equals(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
}
