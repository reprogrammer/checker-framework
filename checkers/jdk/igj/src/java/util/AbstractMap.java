package java.util;

import checkers.igj.quals.*;

@I
public abstract class AbstractMap<K, V> implements @I Map<K, V> {
  @I
  public static class SimpleEntry<K, V> implements @I Map.Entry<K, V>, @I java.io.Serializable {
      private static final long serialVersionUID = 0L;
    public SimpleEntry(K a1, V a2) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public SimpleEntry(@I Map.Entry<? extends K, ? extends V> a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public K getKey() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public V getValue() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public V setValue(V a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public boolean equals(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
    public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public String toString() @ReadOnly { throw new RuntimeException("skeleton method"); }
  }

  @Immutable
  public static class SimpleImmutableEntry<K, V> implements @Immutable Map.Entry<K, V>, @Immutable java.io.Serializable {
      private static final long serialVersionUID = 0L;
    public SimpleImmutableEntry(K a1, V a2) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public SimpleImmutableEntry(Map.Entry<? extends K, ? extends V> a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public K getKey() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public V getValue() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public V setValue(V a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
    public boolean equals(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
    public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
    public String toString() @ReadOnly { throw new RuntimeException("skeleton method"); }
  }

  protected AbstractMap() @ReadOnly {}
  public int size() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean isEmpty() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean containsValue(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean containsKey(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public V get(Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public V put(K a1, V a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public V remove(Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void putAll(@ReadOnly Map<? extends K, ? extends V> a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public @I Set<K> keySet() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I Collection<V> values() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public abstract @I Set<@I Map.Entry<K, V>> entrySet() @ReadOnly;
  public boolean equals(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public String toString() @ReadOnly { throw new RuntimeException("skeleton method"); }
}
