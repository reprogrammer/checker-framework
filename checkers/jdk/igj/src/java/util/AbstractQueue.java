package java.util;
import checkers.igj.quals.*;

@I
public abstract class AbstractQueue<E> extends @I AbstractCollection<E> implements @I Queue<E> {
  protected AbstractQueue() @ReadOnly {}
  public boolean add(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public E remove() @Mutable { throw new RuntimeException("skeleton method"); }
  public E element() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean addAll(@ReadOnly Collection<? extends E> a1) @Mutable { throw new RuntimeException("skeleton method"); }
}
