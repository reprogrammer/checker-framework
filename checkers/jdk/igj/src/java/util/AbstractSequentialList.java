package java.util;
import checkers.igj.quals.*;

@I
public abstract class AbstractSequentialList<E> extends @I AbstractList<E> {
  protected AbstractSequentialList() @ReadOnly {}
  public E get(int a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E set(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public void add(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public E remove(int a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean addAll(int a1, @ReadOnly Collection<? extends E> a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public @I Iterator<E> iterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public abstract @I ListIterator<E> listIterator(int a1) @ReadOnly;
}
