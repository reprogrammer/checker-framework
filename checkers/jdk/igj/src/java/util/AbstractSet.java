package java.util;
import checkers.igj.quals.*;

@I
public abstract class AbstractSet<E> extends @I AbstractCollection<E> implements @I Set<E> {
  protected AbstractSet() @ReadOnly {}
  public boolean equals(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int hashCode() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean removeAll(@ReadOnly Collection<?> a1) @Mutable { throw new RuntimeException("skeleton method"); }
}
