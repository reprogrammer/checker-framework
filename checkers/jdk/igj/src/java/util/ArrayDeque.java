package java.util;
import checkers.igj.quals.*;

@I
public class ArrayDeque<E> extends @I AbstractCollection<E> implements @I Deque<E>, @I Cloneable, @I java.io.Serializable {
    private static final long serialVersionUID = 0L;
  public ArrayDeque() @AssignsFields { throw new RuntimeException("skeleton method"); }
  public ArrayDeque(int a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
  public ArrayDeque(@ReadOnly Collection<? extends E> a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
  public void addFirst(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void addLast(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean offerFirst(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean offerLast(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public E removeFirst() @Mutable { throw new RuntimeException("skeleton method"); }
  public E removeLast() @Mutable { throw new RuntimeException("skeleton method"); }
  public E pollFirst() @Mutable { throw new RuntimeException("skeleton method"); }
  public E pollLast() @Mutable { throw new RuntimeException("skeleton method"); }
  public E getFirst() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E getLast() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E peekFirst() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E peekLast() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean removeFirstOccurrence(@ReadOnly Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean removeLastOccurrence(@ReadOnly Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean add(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean offer(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public E remove() @Mutable { throw new RuntimeException("skeleton method"); }
  public E poll() @Mutable { throw new RuntimeException("skeleton method"); }
  public E element() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E peek() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public void push(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public E pop() @Mutable { throw new RuntimeException("skeleton method"); }
  public int size() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean isEmpty() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I Iterator<E> iterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I Iterator<E> descendingIterator() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean contains(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean remove(@ReadOnly Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public Object[] toArray() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public <T> T[] toArray(T[] a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public @I("N") ArrayDeque<E> clone() { throw new RuntimeException("skeleton method"); }
}
