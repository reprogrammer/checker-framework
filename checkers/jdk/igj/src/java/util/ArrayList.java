package java.util;
import checkers.igj.quals.*;

@I
public class ArrayList<E> extends @I AbstractList<E> implements @I List<E>, @I RandomAccess, @I Cloneable, @I java.io.Serializable {
    private static final long serialVersionUID = 0L;
  public ArrayList(int a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
  public ArrayList() @AssignsFields { throw new RuntimeException("skeleton method"); }
  public ArrayList(@ReadOnly Collection<? extends E> a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
  public void trimToSize() @AssignsFields { throw new RuntimeException("skeleton method"); }
  public void ensureCapacity(int a1) @AssignsFields { throw new RuntimeException("skeleton method"); }
  public int size() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean isEmpty() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public boolean contains(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int indexOf(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public int lastIndexOf(@ReadOnly Object a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public Object[] toArray() @ReadOnly { throw new RuntimeException("skeleton method"); }
  public <T> T[] toArray(T[] a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E get(int a1) @ReadOnly { throw new RuntimeException("skeleton method"); }
  public E set(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean add(E a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void add(int a1, E a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public E remove(int a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean remove(Object a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public void clear() @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean addAll(@ReadOnly Collection<? extends E> a1) @Mutable { throw new RuntimeException("skeleton method"); }
  public boolean addAll(int a1, @ReadOnly Collection<? extends E> a2) @Mutable { throw new RuntimeException("skeleton method"); }
  public @I("N") Object clone() { throw new RuntimeException("skeleton method"); }
}
