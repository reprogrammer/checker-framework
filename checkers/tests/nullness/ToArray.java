import java.util.*;
import checkers.nullness.quals.*;

public class ToArray {
    private List<@Nullable String> nullableList = new ArrayList<@Nullable String>();
    private List<@NonNull String> nonnullList = new ArrayList<@NonNull String>();

    void listToArrayObject() {
        for (@Nullable Object o : nullableList.toArray());
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull Object o : nullableList.toArray());   // error

        for (@Nullable Object o : nonnullList.toArray());
        for (@NonNull Object o : nonnullList.toArray());
    }

    void listToArrayE() {
        for (@Nullable String o : nullableList.toArray(new @Nullable String[0]));
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull String o : nullableList.toArray(new @Nullable String[0]));    // error
        for (@Nullable String o : nullableList.toArray(new @NonNull String[0]));
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull String o : nullableList.toArray(new @NonNull String[0]));

        for (@Nullable String o : nonnullList.toArray(new String[0]));
        // No error expected here. Note that the heuristics determine that the given array
        // is not used and that a new one will be created.
        for (@NonNull String o : nonnullList.toArray(new @Nullable String[0]));
        for (@Nullable String o : nonnullList.toArray(new @NonNull String[0]));
        for (@NonNull String o : nonnullList.toArray(new @NonNull String[0]));
    }

    private Collection<@Nullable String> nullableCol = new ArrayList<@Nullable String>();
    private Collection<@NonNull String> nonnullCol = new ArrayList<@NonNull String>();

    void colToArrayObject() {
        for (@Nullable Object o : nullableCol.toArray());
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull Object o : nullableCol.toArray());    // error

        for (@Nullable Object o : nonnullCol.toArray());
        for (@NonNull Object o : nonnullCol.toArray());
    }

    void colToArrayE() {
        for (@Nullable String o : nullableCol.toArray(new @Nullable String[0]));
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull String o : nullableCol.toArray(new @Nullable String[0])); // error
        for (@Nullable String o : nullableCol.toArray(new @NonNull String[0]));
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull String o : nullableCol.toArray(new @NonNull String[0]));  // error

        for (@Nullable String o : nonnullCol.toArray(new String[0]));
        // No error expected here. Note that the heuristics determine that the given array
        // is not used and that a new one will be created.
        for (@NonNull String o : nonnullCol.toArray(new @Nullable String[0]));
        for (@Nullable String o : nonnullCol.toArray(new @NonNull String[0]));
        for (@NonNull String o : nonnullCol.toArray(new @NonNull String[0]));
    }

    void testHearusitics() {
        for (@Nullable String o : nonnullCol.toArray(new String[] {}));
        for (@NonNull  String o : nonnullCol.toArray(new String[] {}));
        for (@Nullable String o : nonnullCol.toArray(new String[0]));
        for (@NonNull  String o : nonnullCol.toArray(new String[0]));
        for (@Nullable String o : nonnullCol.toArray(new String[nonnullCol.size()]));
        for (@NonNull  String o : nonnullCol.toArray(new String[nonnullCol.size()]));

        for (@Nullable String o : nonnullCol.toArray(new @Nullable String[] {null}));
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull  String o : nonnullCol.toArray(new @Nullable String[] {null})); // error
        for (@Nullable String o : nonnullCol.toArray(new String[1]));
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull  String o : nonnullCol.toArray(new String[1]));   // error
        for (@Nullable String o : nonnullCol.toArray(new String[nonnullCol.size() + 1]));
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull  String o : nonnullCol.toArray(new String[nonnullCol.size() + 1]));   // error

        // cannot handle the following cases for now
        for (@Nullable String o : nonnullCol.toArray(new String[nonnullCol.size() - 1]));
        //:: error: (enhancedfor.type.incompatible)
        for (@NonNull  String o : nonnullCol.toArray(new String[nonnullCol.size() - 1]));   // error
    }
}