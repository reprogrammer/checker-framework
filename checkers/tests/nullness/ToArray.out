:5: (generic.argument.invalid)
:5: (generic.argument.invalid)

:10: (enhancedfor.type.incompatible)
:18: (enhancedfor.type.incompatible)
:20: (enhancedfor.type.incompatible)
:23: (argument.type.incompatible)

:28: (generic.argument.invalid)
:28: (generic.argument.invalid)

:33: (enhancedfor.type.incompatible)
:41: (enhancedfor.type.incompatible)
:43: (enhancedfor.type.incompatible)
:46: (argument.type.incompatible)

:60: (enhancedfor.type.incompatible)
:62: (enhancedfor.type.incompatible)
:64: (enhancedfor.type.incompatible)
:68: (enhancedfor.type.incompatible)
